### ISSUE TITLE
ImageDecoder bahavior different between mixed and cpu mode

### ISSUE THREAD

**Issue by grimoire** [NVIDIA/DALI#1889]
```
import os
import os.path as osp

import nvidia.dali.ops as ops
import nvidia.dali.types as types
from nvidia.dali.pipeline import Pipeline
from nvidia.dali.plugin.pytorch import DALIGenericIterator

import numpy as np
import cv2
import torch

class DaliPipe(Pipeline):
    def __init__(self, batch_size, num_threads, device_id, image_root, image_list_file):
        super(DaliPipe, self).__init__(batch_size, num_threads, device_id)
        self.input = ops.FileReader(file_root=image_root, file_list=image_list_file)
        self.decode = ops.ImageDecoder(device="mixed", output_type=types.BGR)
    
    def define_graph(self):
        self.jpegs, self.label = self.input(name="Reader")
        images = self.decode(self.jpegs)
        return images

def main():
    batch_size = 1
    num_workers = 8
    device = torch.device("cuda:0")
    image_folder = "/home/user/Data/VOC/VOCdevkit/VOC2012/JPEGImages/"
    image_list_file = "./image_list_file.txt"

    print("\n### create dataset")
    # dali
    pipe = DaliPipe(batch_size,
                    num_workers,
                    device.index,
                    image_root=image_folder,
                    image_list_file=image_list_file)

    pipe.build()

    b_dataset_size = int(pipe.epoch_size("Reader"))
    dataloader = DALIGenericIterator(
        pipe, ['data'], dynamic_shape=True, size=b_dataset_size)

    for data in dataloader:
        x = data[0]['data'].to(device)
        torch.cuda.synchronize()
        cv_img = cv2.imread(osp.join(image_folder,'2007_000250.jpg'))
        cv_x = torch.tensor(cv_img).unsqueeze(0).to(device)

        # value different
        x = x.float()
        cv_x = cv_x.float()
        diff = (x-cv_x).abs()
        print("diff max:",diff.max())
        print("diff mean:",diff.mean())

        # different visualize
        img = x.cpu().squeeze().numpy()
        img = img.astype(float)-cv_img.astype(float)
        img = np.abs(img).astype(np.uint8)*10
        
        img = np.concatenate([img, cv_img], 1)
        cv2.imshow('img',img)
        cv2.waitKey()
        cv2.destroyAllWindows()


if __name__ == "__main__":
    main()
```

there is only one image (2007_000250.jpg) in image_list_file.txt, with a random label

when I load  the image with mode "mixed", the result image is differet from the one read by opencv.
![1](https://user-images.githubusercontent.com/1239736/79973517-47bc3b80-84ca-11ea-968f-724103f72087.jpg)

If i use "cpu" mode instead, there will be no different.
![2](https://user-images.githubusercontent.com/1239736/79973565-5dc9fc00-84ca-11ea-838d-0934f296c163.jpg)

most difference happen on edge pixels. Is the any parameter can fix this?
Thank

**Comment by JanuszL** [comment]
Hi,
This is expected and both images are properly decoded.
For the CPU and OpenCV, you have a match as we are using libjpeg-turbo library in DALI and OpenCV is using it as well under the hood for the JPEG decoding. For the GPU we are using nvJPEG.
There are two sources of difference:
- idct
- color conversion
nvJPEG should be in the range of ~3-pixel values comparing to libjpeg. However, libjpeg-turbo by default is using something called "fancy-upsampling" which makes this difference bigger in your test.
JPEG standard doesn't define any gold standard for decoding and some difference is allowed. You can read more about libjpeg-turbo approach [here](https://github.com/libjpeg-turbo/libjpeg-turbo#mathematical-compatibility).

**Comment by grimoire** [comment]
> Hi,
> This is expected and both images are properly decoded.
> For the CPU and OpenCV, you have a match as we are using libjpeg-turbo library in DALI and OpenCV is using it as well under the hood for the JPEG decoding. For the GPU we are using nvJPEG.
> There are two sources of difference:
> 
> * idct
> * color conversion
>   nvJPEG should be in the range of ~3-pixel values comparing to libjpeg. However, libjpeg-turbo by default is using something called "fancy-upsampling" which makes this difference bigger in your test.
>   JPEG standard doesn't define any gold standard for decoding and some difference is allowed. You can read more about libjpeg-turbo approach [here](https://github.com/libjpeg-turbo/libjpeg-turbo#mathematical-compatibility).

thanks for reply!
I also found that some images can be decoded by opencv but failed in nvjpeg (Premature or file read error).  Guess it is caused by this too.

by the way,
My old detection model is trained with opencv decoder, when testing on nvjpeg decoder, It give me 0.7 map drop. Especially on crack detection. 
I want to use the libjpeg-turbo as the decoder, but rest of the pipeline with gpu version(resize, normalize,...etc), is it possible?

**Comment by JanuszL** [comment]
> I also found that some images can be decoded by opencv but failed in nvjpeg (Premature or file read error). Guess it is caused by this too.

This is a different problem. Probably data is corrupted and OpenCV is more forgettable about the format, while DALI doesn't have elaborate heuristics to correct invalid image headers. If you can share any image that is causing problems please share it and we will check it.

> My old detection model is trained with opencv decoder, when testing on nvjpeg decoder, It give me 0.7 map drop. Especially on crack detection.

It is expected as you have a slightly different input to your network. In real-life your samples will vary anyway so this won't be a problem, but for testing, you should use the same pipeline configuration for training as for validation.

> I want to use the libjpeg-turbo as the decoder, but rest of the pipeline with gpu version(resize, normalize,...etc), is it possible?

It is, just:
```
def __init__(self, ...):
        super(MyPipeline, self).__init__(...)
        self.input = ops.FileReader(file_root=data_dir, shard_id=args.local_rank, num_shards=args.world_size, random_shuffle=True)
        self.decode = ops.ImageDecoder(device="cpu", output_type=types.RGB)

    def define_graph(self):
        self.jpegs, self.labels = self.input(name="Reader")
        images = self.decode(self.jpegs)
        images = self.some_op(images.gpu())
        (...)
```

**Comment by grimoire** [comment]
> 
> This is a different problem. Probably data is corrupted and OpenCV is more forgettable about the format, while DALI doesn't have elaborate heuristics to correct invalid image headers. If you can share any image that is causing problems please share it and we will check it.
> 

thanks for the advice.

here are two images that cause the problem

[tmp.zip](https://github.com/NVIDIA/DALI/files/4521265/tmp.zip)

**Comment by JanuszL** [comment]
I see that both images are corrupted what is in line with the error message I see in my debug build, "dali/dali/image/jpeg_mem.cc:338: Premature end of JPEG data. Stopped at line 2830/2976"
 ![image](https://user-images.githubusercontent.com/39967756/80080368-cdcf9500-8551-11ea-9ed3-8a446d9277a8.png).
Also, the "mixed" decoder falls back to the CPU backend anyway (nvJPEG cannot handle this probably because of this corruption).

**Comment by grimoire** [comment]
Hi, 
Sorry to bother you again.
If different jpeg decoder generate different images, is it possible to do some data augmentation with decoder, such as change IDCT/ quanti table/ upsample etc ?  Will it be different from image blur/ denoise or other common augmentation method? 
Just curious.

**Comment by JanuszL** [comment]
HI,
I'm not that much into JPEGs after all. What JPEG decoding libraries can do is mostly allow you a fast down-sample using 2, 4, or 8 scales as the basic block that JPEG operates on is 8x8 size, so you can decode only part of it what should save a bit of computation without any loss of the quality. Other operations you mentioned are up tot the decoder library implementation. So theoretically it is possible but I don't know if libraries like libjpeg-turbo expose it in its the API.
